### PR TITLE
MAISTRA-2589 Do not allow 2.1 control planes with Mixer enabled

### PR DESCRIPTION
**Pull request by dgn**
This will make sure people cannot upgrade to 2.1 when they have Mixer enabled.

### CODE REVIEW

**Review by jwendell — APPROVED** [review]

File: pkg/controller/versions/strategy_v2_1.go [hunk 1]
```diff
@@ -142,6 +143,16 @@ func (v *versionStrategyV2_1) validateProtocolDetection(ctx context.Context, met
 	return allErrors
 }
 
+func (v *versionStrategyV2_1) validateMixerDisabled(spec *v2.ControlPlaneSpec, allErrors []error) []error {
+	if spec.Policy != nil && (spec.Policy.Type == v2.PolicyTypeMixer || spec.Policy.Mixer != nil) {
+		allErrors = append(allErrors, fmt.Errorf("support for policy.type '%s' and policy.Mixer options been removed in v2.1, please use another alternative", v2.PolicyTypeMixer))
```
Comment (shamsher31) [inline pkg/controller/versions/strategy_v2_1.go]: nit:
```suggestion
		allErrors = append(allErrors, fmt.Errorf("support for policy.type %q and policy.Mixer options have been removed in v2.1, please use another alternative", v2.PolicyTypeMixer))
```

File: pkg/controller/versions/strategy_v2_1.go [hunk 2]
```diff
@@ -142,6 +143,16 @@ func (v *versionStrategyV2_1) validateProtocolDetection(ctx context.Context, met
...
+	}
+	if spec.Telemetry != nil && (spec.Telemetry.Type == v2.TelemetryTypeMixer || spec.Telemetry.Mixer != nil) {
+		allErrors = append(allErrors, fmt.Errorf("support for telemetry.type '%s' and telemetry.Mixer options have been removed in v2.1, please use another alternative", v2.TelemetryTypeMixer))
```
Comment (shamsher31) [inline pkg/controller/versions/strategy_v2_1.go]:
```suggestion
		allErrors = append(allErrors, fmt.Errorf("support for telemetry.type %q and telemetry.Mixer options have been removed in v2.1, please use another alternative", v2.TelemetryTypeMixer))
```

**Review by shamsher31 — APPROVED** [review]

**Review by rcernich — APPROVED** [review]
Should we also scan for config.istio.io resources, or is it good enough to ensure mixer is disabled?  I assume, if mixer is disabled, then any handler/instance resources will simply be ignored if mixer isn't present, but I'm not sure.  The goal is to ensure that existing sidecars won't be affected when disabling mixer (i.e. if mixer goes away, do they need to be reinjected?).  That said, if this is an issue, then it probably exists in 2.0 too, in which case we can create a follow-up issue to deal with that.  (Maybe it's good enough to "fail open" when mixer is removed.)

**Comment by dgn** [comment]
> Should we also scan for config.istio.io resources, or is it good enough to ensure mixer is disabled? I assume, if mixer is disabled, then any handler/instance resources will simply be ignored if mixer isn't present, but I'm not sure. The goal is to ensure that existing sidecars won't be affected when disabling mixer (i.e. if mixer goes away, do they need to be reinjected?). That said, if this is an issue, then it probably exists in 2.0 too, in which case we can create a follow-up issue to deal with that. (Maybe it's good enough to "fail open" when mixer is removed.)

I think generally we should try to avoid looking at resources outside the SMCP when validating it, simply for performance reasons. If customers have handler/instance resources around, but Mixer disabled, they cannot expect them to work anyway

**Pull request merged**